File: backend/skills/locations.go
package skills

import (
	"os"
	"path/filepath"
	"strings"

	"github.com/enough/enough/backend/config"
	"github.com/enough/enough/backend/enoughhome"
)

type SearchDir struct {
	Path          string
	Source        string // "project" | "user" | "path" | "flame"
	IncludeRootMD bool   // false for .agents/skills only
}

func isBlockedSkillRoot(path string) bool {
	clean := filepath.Clean(path)
	return strings.Contains(clean, filepath.Join(".cursor", "skills-cursor"))
}

func hasExcludedComponent(path string) bool {
	clean := filepath.Clean(path)
	parts := strings.Split(clean, string(filepath.Separator))
	for _, part := range parts {
		if ExcludedSkillDirs[part] {
			return true
		}
	}
	return false
}

// SearchLocations returns every directory to scan for skills, in STRICT
// precedence order (first match wins on name collision).
func SearchLocations(workDir string, cfg config.Runtime, agentDirOverride string) []SearchDir {
	var dirs []SearchDir
	seen := make(map[string]bool)

	addDir := func(path, source string, includeRootMD bool) {
		abs, err := filepath.Abs(path)
		if err != nil {
			abs = filepath.Clean(path)
		}
		if seen[abs] {
			return
		}
		if isBlockedSkillRoot(abs) || hasExcludedComponent(abs) {
			return
		}
		seen[abs] = true
		dirs = append(dirs, SearchDir{
			Path:          abs,
			Source:        source,
			IncludeRootMD: includeRootMD,
		})
	}

	// 1 & 2. Project: cwd→gitRoot (or FS root if no git repo)
	current := workDir
	gitRoot := ""
	for {
		if _, err := os.Stat(filepath.Join(current, ".git")); err == nil {
			gitRoot = current
			break
		}
		parent := filepath.Dir(current)
		if parent == current {
			break
		}
		current = parent
	}

	current = workDir
	for {
		// .enough/skills
		addDir(filepath.Join(current, ".enough", "skills"), "project", true)
		// .agents/skills
		addDir(filepath.Join(current, ".agents", "skills"), "project", false)

		if gitRoot != "" && current == gitRoot {
			break
		}
		parent := filepath.Dir(current)
		if parent == current {
			break
		}
		current = parent
	}

	// cwd .cursor/skills
	addDir(filepath.Join(workDir, ".cursor", "skills"), "project", true)

	// 3. Global user: ~/.enough/skills, ~/.enough/agent/skills, ~/.agents/skills, ~/.cursor/skills
	home := enoughhome.HomeDir()
	addDir(filepath.Join(home, "skills"), "user", true)
	if agentDirOverride != "" {
		addDir(filepath.Join(agentDirOverride, "skills"), "user", true)
	} else {
		addDir(filepath.Join(home, "agent", "skills"), "user", true)
	}

	userHome, err := os.UserHomeDir()
	if err == nil {
		addDir(filepath.Join(userHome, ".agents", "skills"), "user", false)
		addDir(filepath.Join(userHome, ".cursor", "skills"), "user", true)
	}

	// 4. Optional read-only: ~/.flame/skills (if exists)
	if userHome != "" {
		flameSkills := filepath.Join(userHome, ".flame", "skills")
		if fi, err := os.Stat(flameSkills); err == nil && fi.IsDir() {
			addDir(flameSkills, "flame", true)
		}
	}

	// 5. Config explicit paths (cfg.Skills.Paths, minus ! exclusions)
	for _, p := range cfg.Skills.Paths {
		if strings.HasPrefix(p, "!") {
			continue
		}
		absPath := p
		if !filepath.IsAbs(absPath) {
			if strings.HasPrefix(absPath, "~") && userHome != "" {
				absPath = filepath.Join(userHome, absPath[1:])
			} else {
				absPath = filepath.Join(workDir, absPath)
			}
		}
		addDir(absPath, "path", true)
	}

	// 6. External directories from config (Hermes semantics)
	for _, extDir := range getExternalSkillsDirs(cfg) {
		addDir(extDir, "user", true)
	}

	return dirs
}

func getExternalSkillsDirs(cfg config.Runtime) []string {
	userHome, _ := os.UserHomeDir()
	home := enoughhome.HomeDir()
	localSkills, _ := filepath.Abs(filepath.Join(home, "skills"))

	var out []string
	seen := make(map[string]bool)

	for _, entry := range cfg.Skills.ExternalDirs {
		entry = strings.TrimSpace(entry)
		if entry == "" {
			continue
		}

		// Expand env variables
		expanded := os.ExpandEnv(entry)
		// Expand ~
		if strings.HasPrefix(expanded, "~") && userHome != "" {
			expanded = filepath.Join(userHome, expanded[1:])
		}

		// Resolve relative paths against ENOUGH_HOME (home)
		abs := expanded
		if !filepath.IsAbs(abs) {
			abs = filepath.Join(home, abs)
		}
		abs, err := filepath.Abs(abs)
		if err != nil {
			abs = filepath.Clean(abs)
		}

		if abs == localSkills {
			continue
		}
		if seen[abs] {
			continue
		}

		fi, err := os.Stat(abs)
		if err == nil && fi.IsDir() {
			seen[abs] = true
			out = append(out, abs)
		}
	}
	return out
}

File: backend/skills/locations.ts
// PORT: backend/skills/locations.go

import fs from "node:fs";
import path from "node:path";
import os from "node:os";
import { type runtime } from "../config/config";
import { HomeDir } from "./paths";
import { ExcludedSkillDirs } from "./constants";

export interface SearchDir {
  Path: string;
  Source: string; // "project" | "user" | "path" | "flame"
  IncludeRootMD: boolean; // false for .agents/skills only
}

function isBlockedSkillRoot(filePath: string): boolean {
  const clean = path.normalize(filePath);
  return clean.includes(path.join(".cursor", "skills-cursor"));
}

function hasExcludedComponent(filePath: string): boolean {
  const clean = path.normalize(filePath);
  const parts = clean.split(path.sep);
  for (const part of parts) {
    if (ExcludedSkillDirs[part]) {
      return true;
    }
  }
  return false;
}

// SearchLocations returns every directory to scan for skills, in STRICT
// precedence order (first match wins on name collision).
export function SearchLocations(workDir: string, cfg: runtime, agentDirOverride: string): SearchDir[] {
  const dirs: SearchDir[] = [];
  const seen = new Set<string>();

  const addDir = (dirPath: string, source: string, includeRootMD: boolean) => {
    let abs: string;
    try {
      abs = path.resolve(dirPath);
    } catch {
      abs = path.normalize(dirPath);
    }
    if (seen.has(abs)) {
      return;
    }
    if (isBlockedSkillRoot(abs) || hasExcludedComponent(abs)) {
      return;
    }
    seen.add(abs);
    dirs.push({
      Path: abs,
      Source: source,
      IncludeRootMD: includeRootMD,
    });
  };

  // 1 & 2. Project: cwd→gitRoot (or FS root if no git repo)
  let current = workDir;
  let gitRoot = "";
  while (true) {
    if (fs.existsSync(path.join(current, ".git"))) {
      gitRoot = current;
      break;
    }
    const parent = path.dirname(current);
    if (parent === current) {
      break;
    }
    current = parent;
  }

  current = workDir;
  while (true) {
    // .hollow/skills
    addDir(path.join(current, ".hollow", "skills"), "project", true);
    // .agents/skills
    addDir(path.join(current, ".agents", "skills"), "project", false);

    if (gitRoot !== "" && current === gitRoot) {
      break;
    }
    const parent = path.dirname(current);
    if (parent === current) {
      break;
    }
    current = parent;
  }

  // cwd .cursor/skills
  addDir(path.join(workDir, ".cursor", "skills"), "project", true);

  // 3. Global user: ~/.hollow/skills, ~/.hollow/agent/skills, ~/.agents/skills, ~/.cursor/skills
  const home = HomeDir();
  addDir(path.join(home, "skills"), "user", true);
  if (agentDirOverride !== "") {
    addDir(path.join(agentDirOverride, "skills"), "user", true);
  } else {
    addDir(path.join(home, "agent", "skills"), "user", true);
  }

  const userHome = os.homedir();
  if (userHome) {
    addDir(path.join(userHome, ".agents", "skills"), "user", false);
    addDir(path.join(userHome, ".cursor", "skills"), "user", true);
  }

  // 4. Optional read-only: ~/.flame/skills (if exists)
  if (userHome) {
    const flameSkills = path.join(userHome, ".flame", "skills");
    try {
      const fi = fs.statSync(flameSkills);
      if (fi.isDirectory()) {
        addDir(flameSkills, "flame", true);
      }
    } catch {}
  }

  // 5. Config explicit paths (cfg.skills.paths, minus ! exclusions)
  const pathsList = cfg.skills?.paths || [];
  for (const p of pathsList) {
    if (p.startsWith("!")) {
      continue;
    }
    let absPath = p;
    if (!path.isAbsolute(absPath)) {
      if (absPath.startsWith("~") && userHome) {
        absPath = path.join(userHome, absPath.substring(1));
      } else {
        absPath = path.join(workDir, absPath);
      }
    }
    addDir(absPath, "path", true);
  }

  // 6. External directories from config (Hermes semantics)
  for (const extDir of getExternalSkillsDirs(cfg)) {
    addDir(extDir, "user", true);
  }

  return dirs;
}

function getExternalSkillsDirs(cfg: runtime): string[] {
  const userHome = os.homedir();
  const home = HomeDir();
  let localSkills: string;
  try {
    localSkills = path.resolve(path.join(home, "skills"));
  } catch {
    localSkills = path.normalize(path.join(home, "skills"));
  }

  const out: string[] = [];
  const seen = new Set<string>();
  const extDirs = cfg.skills?.external_dirs || [];

  for (let entry of extDirs) {
    entry = entry.trim();
    if (entry === "") {
      continue;
    }

    // Expand env variables
    let expanded = entry.replace(/\$(\w+)|\${(\w+)}/g, (_, m1, m2) => process.env[m1 || m2] || "");

    // Expand ~
    if (expanded.startsWith("~") && userHome) {
      expanded = path.join(userHome, expanded.substring(1));
    }

    // Resolve relative paths against HOLLOW_HOME (home)
    let abs = expanded;
    if (!path.isAbsolute(abs)) {
      abs = path.join(home, abs);
    }
    try {
      abs = path.resolve(abs);
    } catch {
      abs = path.normalize(abs);
    }

    if (abs === localSkills) {
      continue;
    }
    if (seen.has(abs)) {
      continue;
    }

    try {
      const fi = fs.statSync(abs);
      if (fi.isDirectory()) {
        seen.add(abs);
        out.push(abs);
      }
    } catch {}
  }
  return out;
}

/*
PORT STATUS
source path: backend/skills/locations.go
source lines: 187
draft lines: 185
confidence: high
status: phase_b_compile
*/
